Show source by default, add feature to hide it

schneems opened the pull request:
close #59

Changes the default behavior so that the output in the readme now matches the output a user would see when they use the library. Specifically, it now includes the original `std::io::Error` in the output.

It introduces a feature `anyhow`. By default:

- By default: fs-err will include `std::io::Error` in the Display output, and return `None` from `Error::source()`
- With `anyhow` fs-err will not include std::io::Error in the Display output, and return `Some(std::io::Error)` from `Error::source()`

This is based on the guidance from #51. That discussion links to this 2020 discussion https://github.com/rust-lang/project-error-handling/issues/23 that suggests that you should either print the source and return `None` from `Error::source()` (https://doc.rust-lang.org/std/error/trait.Error.html#method.source) or not print anything and return `Some(E)`.

This allows users of anyhow (or similar) libraries to configure fs-err for the behavior they desire, while not hiding information by default from unsuspecting adopters that are not using those libraries. It optimizes for the case of accidentally showing extra information that a user can then investigate and disable, rather than hiding information that the user might not realize is missing.

andrewhickman commented:
I'm leaning towards of `expose_original_error` for the feature flag name. I imagine it would mainly be used when you need direct access to the original std error for some reason. Users who only want to print the error shouldn't need to care about whether it is enabled or not.

schneems commented:
I think that's a pragmatic choice. I updated the PR with that wording. My only concern would be someone mistaking the intent when developing internally. To that end, I added a comment explaining it inside of the `Display` implementations. I don't think it's strictly required if you don't like it. Commits are small and layered on top of prior work.

schneems commented on src/errors.rs:
```diff
@@ -188,7 +199,12 @@ impl fmt::Display for SourceDestError {
+        }?;
+
+        #[cfg(not(feature = "custom_caused_by"))]
+        write!(formatter, "    caused by: {}", self.source)?;
```
Sorry. I didn't see this comment until after it was merged. I've been checking the PR. I see it was made 4 days ago, maybe it was in a pending review state or maybe github had a hiccup?

Either way. I also see you merged it and made the change yourself, which is perfectly fine with me. Thanks again
